### PR TITLE
Fixed a bug in pileup mode relating to number of haplotypes

### PR DESCRIPTION
**Pull request by jonn-smith**
Recent refactoring seems to have introduced a bug in pileup mode that failed to enforce the limit on the number of haplotypes to be considered.  With this patch:

- HaplotypeCaller once again respects the limit on haplotypes before genotyping.
- Changed some `HashSet`s to `LinkedHashSets` to preserve determinism.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java [hunk 1]
```diff
@@ -470,12 +470,21 @@ static Set<Haplotype> filterPileupHaplotypes(final Set<Haplotype> onlyNewHaploty
...
+        final long minimumPassingScore = scores.values().stream()
+                .sorted(Comparator.reverseOrder())
+                .skip(numPileupHaplotypes - 1)
+                .findFirst().get();
+
+        // If there are ties, we pass any haplotype with a score as good as the numPileupHaplotypes-th best, with
+        // final ordering determined by string representation (for determinism).
+        return onlyNewHaplotypes.stream()
+                .filter(h -> scores.get(h) >= minimumPassingScore)
+                .sorted(Comparator.comparing(Object::toString))
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:483]: Haplotuype::getBaseString()

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtilsUnitTest.java [hunk 2]
```diff
@@ -1312,25 +1312,28 @@ public Object[][] filterPileupHaplotypesDataProvider() {
...
-                new Object[]{Arrays.asList(hapA,hapB,hapD,hapF),hapFRepeatedKmers,1,3,Arrays.asList(hapF, hapD)}, //currently repeated kmers only count as singular evidence
+                new Object[]{Arrays.asList(hapA,hapB,hapD,hapF),hapFRepeatedKmers,2,3,Arrays.asList(hapF, hapD)},
+                new Object[]{Arrays.asList(hapA,hapB,hapD,hapF),hapFRepeatedKmers,1,3, List.of(hapD) }, //currently repeated kmers only count as singular evidence
 
                 // These tests demonstrate that the weights in the map don't matter
-                new Object[]{Arrays.asList(hapA,hapB,hapC,hapD),hapDKmersHighSupport,1,3,Arrays.asList(hapA,hapB,hapC,hapD)},
-                new Object[]{Arrays.asList(hapA,hapB,hapC,hapD),hapDKmersHighSupport,2,3,Arrays.asList(hapA,hapB,hapC,hapD)},
-                new Object[]{Arrays.asList(hapA,hapB,hapC,hapD),hapDKmersHighSupport,3,3,Arrays.asList(hapA,hapB,hapC,hapD)}, // Despite hapD having good support it is not weighted higher
+                new Object[]{Arrays.asList(hapA,hapB,hapC,hapD),hapDKmersHighSupport,1,3, List.of(hapD) },
```
Comment (jamesemery) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtilsUnitTest.java:1328]: Update the comments with an explanation that now numPileupHaplotypes overrides the score for tiebreaking... 

Also i think a test illustrating this fix where there are a dozen haplotypes (copying them here is fine) with the same score and asserting that you will still get numHaplotypes back would be best for this.

File: src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf [hunk 3]
```diff
@@ -145,11 +145,8 @@
...
-20	10099046	.	T	C	73.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=-1.822;DP=18;ExcessHet=0.0000;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=55.26;MQRankSum=-3.419;QD=4.09;ReadPosRankSum=-1.364;SOR=0.193	GT:AD:DP:GQ:PL	0/1:15,3:18:81:81,0,561
-20	10099055	.	T	C	47.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=-1.460;DP=19;ExcessHet=0.0000;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=55.52;MQRankSum=-3.419;QD=2.65;ReadPosRankSum=-0.830;SOR=0.330	GT:AD:DP:GQ:PL	0/1:15,3:18:55:55,0,557
-20	10099079	.	C	T	40.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=0.694;DP=23;ExcessHet=0.0000;FS=2.192;MLEAC=1;MLEAF=0.500;MQ=56.33;MQRankSum=-3.254;QD=1.77;ReadPosRankSum=-0.568;SOR=1.460	GT:AD:DP:GQ:PL	0/1:19,4:23:48:48,0,551
-20	10099140	.	G	T	583.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=-0.176;DP=49;ExcessHet=0.0000;FS=6.794;MLEAC=1;MLEAF=0.500;MQ=60.07;MQRankSum=1.344;QD=12.42;ReadPosRankSum=-0.466;SOR=1.944	GT:AD:DP:GQ:PL	0/1:28,19:47:99:591,0,932
+20	10099046	.	T	C	31.64	.	AC=1;AF=0.500;AN=2;BaseQRankSum=-1.842;DP=17;ExcessHet=0.0000;FS=3.701;MLEAC=1;MLEAF=0.500;MQ=56.07;MQRankSum=-3.254;QD=1.86;ReadPosRankSum=-0.597;SOR=0.090	GT:AD:DP:GQ:PL	0/1:15,2:17:39:39,0,564
```
Comment (jamesemery) [inline src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: these diferences are mostly noise to do with exactly what haplotypes come out... it is interesting the ties at the 5th score level seem to be very common... We really need to replace this heuristic with something not quite so terrible... @davidbenjamin

**Review by jamesemery — CHANGES_REQUESTED** [review]
One minor change and a request to update the unit tests to reflect the new reality a little better. 

@davidbenjamin This is a reversion to a fix you made recently fixing the non-determninism... it turns out this was in some edge cases producing assembly windows with 50k+ haplotypes (because we were taking everything that ties at the 5 ranked score without limit...) Now we are arbitrarily breaking based on string comparision which at least prevents us from exploding... 
 
Do you have any ideas for an alternative tie-breaking scheme for haploytpes? It seems these ties were happening much more than expected (probably because most haplotypes end up with (length - k) as their scores because there kmers appear at least somewhere... This also would seem to bias us to mostly select the longest haplotypes in every case for inserting events....

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java (hunk 1 above)
Comment (jonn-smith) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:483]: 👍

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtilsUnitTest.java (hunk 2 above)
Comment (jonn-smith) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtilsUnitTest.java:1328]: I added a short comment on line 1317, but I'll make it more explicit.

I'll add another test as you suggest.

File: src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf (hunk 3 above)
Comment (davidbenjamin) [reply src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: @jamesemery It seems to be that the intent here is to corral the combinatorial explosion of paths in the de Bruijn graph that don't correspond to actual phasing.  However, by using the same kmer size as in assembly the heuristic is completely toothless!  It's basically a tautology that all the kmers of an assembled haplotype show up in some read, since otherwise the kmer wouldn't exist in the graph.

A better approach would be to use the same heuristic but with larger kmers than used in assembly.  For example, using k = 30 would get rid of spurious phasing within 30 bases.

Also, I don't think that using a larger k is merely a tie-breaker for the extant approach.  Rather, what we have now  should be replaced completely.

I don't think I am done pondering this but that's a good start.

Comment (davidbenjamin) [reply src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: One might also ask whether an excess of haplotypes means that instead of filtering we should just redo assembly with a larger k, or dynamically switch to linked de Bruijn graphs.  That could well be.

Comment (jamesemery) [reply src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: That seems like a good observation... It would also make it harder for every haplotype to be maximally represented by reads! I think it might also be worth considering normalizing to the lenght of the haplotype as well to prevent ourselves from being biased towards longer haplotypes that have incidental (and distant) insertions being over-represented in the choices here... alternatively we could flip the score on its head... where the score is the absolute number of missing kmers from the reads, which should more or less be capped at ~(2k -1) but not have this gross property that it prefers longer hapltotypes

Comment (davidbenjamin) [reply src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: I agree that counting non-missing kmers and thereby rewarding length is wrong.

Since pileup haplotypes are constructed by injecting pileup events into base haplotypes, why not compute the score at that point as follows: 1) make a single kmer containing the pileup event and any nearby events on the base haplotype 2) check whether this kmer is supported in the reads.  If not, it's a spurious phasing. 3) If too many pileup haplotypes pass, repeat with larger k.

Comment (jamesemery) [reply src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: @davidbenjamin i'm going to open a ticket to re-visit this heuristic... i think the band-aid here is sufficient and necessary to get into the next release.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java [hunk 4]
```diff
@@ -470,12 +470,21 @@ static Set<Haplotype> filterPileupHaplotypes(final Set<Haplotype> onlyNewHaploty
...
+        final long minimumPassingScore = scores.values().stream()
+                .sorted(Comparator.reverseOrder())
+                .skip(numPileupHaplotypes - 1)
+                .findFirst().get();
+
+        // If there are ties, we pass any haplotype with a score as good as the numPileupHaplotypes-th best, with
+        // final ordering determined by string representation (for determinism).
+        return onlyNewHaplotypes.stream()
+                .filter(h -> scores.get(h) >= minimumPassingScore)
+                .sorted(Comparator.comparing(Haplotype::getBaseString))
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java:483]: Actually i think you need to change this... you need Comparitor.comparing(score).thenComparing (need to check the syntax on that) because now this is shuffling the score > minPassingScore haplotypes which is probably fine most of the time but could end up not selecting the top scorign hap.

**Review by jamesemery — COMMENTED** [review]
One last minute fix then i think this can be merged

File: src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf (hunk 3 above)
Comment (davidbenjamin) [reply src/test/resources/org/broadinstitute/hellbender/tools/haplotypecaller/expected.pileupCallerDRAGEN.gatk4.vcf:148]: Agreed.

**Review by jamesemery — APPROVED** [review]
Once you fix that subtlety in the comparator feel free to merge this branch.
